@here/harp-mapview/test/TextElementsRendererTest.ts

```diff
@@ -0,0 +1,279 @@
+        });
+
+        const renderedTextElements: TextElement[] = [];
+        const secondChanceTextElements: TextElement[] = [];
+
+        // time must not be 0 b/c 0 is used as a special value in TextElementsRenderer.
+        const time = 1000;
+        const frameNumber = 0;
+        const zoomLevel = 4;
+
+        // TextElementRenderer.initializeTextCanvases is triggering a mapView.update
+        // Wait for that update call otherwise we will not see any text.
+        updatePromise = new Promise((resolve, reject) => {
+            mapView.update = resolve;
+        });
+        let numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time,
+            frameNumber,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+
+        // First call just triggers the text canvas initialization. Nothing will be rendered.
+        expect(numRenderedTextElements).to.be.equal(0);
+        expect(renderedTextElements.length).to.be.equal(0);
+
+        // TextElementsRenderer.renderTextElements is loading the charset lazily
+        // (FontCatalog.loadCharset) and triggers a mapView.update afterwards.
+        await updatePromise;
+
+        // First "real" frame. Label should be invisble and start fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 100,
+            frameNumber + 1,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+        expect(numRenderedTextElements).to.be.equal(1);
+        expect(renderedTextElements.length).to.be.equal(1);
+        expect(renderedTextElements[0].textRenderState!.opacity).to.be.equal(0);
+        assert(renderedTextElements[0].textRenderState!.isFadingIn());
+
+        // Second frame. Label should be semi-transparent and fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 200,
+            frameNumber + 2,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+
+        expect(numRenderedTextElements).to.be.equal(1);
+        expect(renderedTextElements.length).to.be.equal(1);
+        expect(renderedTextElements[0].textRenderState!.opacity)
+            .to.be.greaterThan(0)
+            .to.be.lessThan(1);
+        assert(renderedTextElements[0].textRenderState!.isFadingIn());
+
+        const lastOpacity = renderedTextElements[0].textRenderState!.opacity;
+
+        // Third frame. Label should be semi-transparent and still fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 800,
+            frameNumber + 8,
```

Comment:
should this be +3?

Reply:
True. I actually don't know if the frameNumber is used for anything. I think the only requirement is, that it is monotonically increasing